File: src/main/java/committee/nova/ns/common/util/Utilities.java
package committee.nova.ns.common.util;

import net.minecraft.World;
import net.minecraft.block.*;

import java.util.Random;

public class Utilities {
    public static boolean tryCatalyze(World world, double[] pos) {
        final int i = (int) Math.floor(pos[0]);
        final int j = (int) Math.floor(pos[1]);
        final int k = (int) Math.floor(pos[2]);
        final int var8 = world.getBlockId(i, j, k);
        if (var8 == Block.SAPLING.id) {
            if (!world.isClient) ((SaplingBlock) Block.SAPLING).method_2221(world, i, j, k, world.random);
            return true;
        } else if (var8 != Block.BROWN_MUSHROOM.id && var8 != Block.RED_MUSHROOM.id) {
            if (var8 != Block.MELON_STEM.id && var8 != Block.PUMPKIN_STEM.id) {
                if (var8 == Block.WHEAT_CROP.id) {
                    if (!world.isClient) ((CropBlock) Block.WHEAT_CROP).method_2389(world, i, j, k);
                    return true;
                } else if (var8 == Block.GRASS.id) {
                    if (!world.isClient) {
                        label73:
                        for (int var9 = 0; var9 < 128; ++var9) {
                            int var10 = i;
                            int var11 = j + 1;
                            int var12 = k;
                            final Random random = world.random;
                            for (int var13 = 0; var13 < var9 / 16; ++var13) {
                                var10 += random.nextInt(3) - 1;
                                var11 += (random.nextInt(3) - 1) * random.nextInt(3) / 2;
                                var12 += random.nextInt(3) - 1;
                                if (world.getBlockId(var10, var11 - 1, var12) != Block.GRASS.id || world.canSuffocate(var10, var11, var12)) {
                                    continue label73;
                                }
                            }

                            if (world.getBlockId(var10, var11, var12) == 0) {
                                if (random.nextInt(10) != 0) {
                                    world.notify(var10, var11, var12, Block.TALL_GRASS.id, 1);
                                } else if (random.nextInt(3) != 0) {
                                    world.setBlock(var10, var11, var12, Block.FLOWER.id);
                                } else {
                                    world.setBlock(var10, var11, var12, Block.ROSE.id);
                                }
                            }
                        }
                    }

                    return true;
                } else {
                    return false;
                }
            } else {
                if (!world.isClient) ((StemBlock) Block.BY_ID[var8]).method_639(world, i, j, k);
                return true;
            }
        } else {
            if (!world.isClient)
                return ((MushroomPlantBlock) Block.BY_ID[var8]).method_827(world, i, j, k, world.random);
            return true;
        }
    }
}

File: src/main/scala/committee/nova/ns/common/util/Utilities.scala
package committee.nova.ns.common.util

import committee.nova.ns.common.config.CommonConfig._
import net.minecraft.block._
import net.minecraft.entity.item.EntityItem
import net.minecraft.world.World
import net.minecraftforge.common.ForgeHooks

import scala.util.Random
import scala.util.control.Breaks._

object Utilities {
  def catalyze(entity: EntityItem): Boolean = {
    val world = entity.worldObj
    val x = Math.floor(entity.posX).intValue()
    val y = Math.floor(entity.posY + 0.2).intValue()
    val z = Math.floor(entity.posZ).intValue()
    val blockIn = world.getBlockId(x, y, z)
    if (influencedByBiome && !willTryCatalyze(world, x, z)) return false
    if (tryCatalyze(world, blockIn, x, y, z)) return true
    val dirt = world.getBlockId(x, y - 1, z)
    tryCatalyze(world, dirt, x, y - 1, z)
  }

  def tryCatalyze(world: World, block: Int, x: Int, y: Int, z: Int): Boolean = {
    if (block == Block.sapling.blockID) {
      if (!world.isRemote) {
        if (world.rand.nextFloat.toDouble < 0.45D) Block.sapling.asInstanceOf[BlockSapling].markOrGrowMarked(world, x, y, z, world.rand)
      }
      true
    }
    else if (block != Block.mushroomBrown.blockID && block != Block.mushroomRed.blockID) if (block != Block.melonStem.blockID && block != Block.pumpkinStem.blockID) if (block > 0 && Block.blocksList(block).isInstanceOf[BlockCrops]) if (world.getBlockMetadata(x, y, z) == 7) return false
    else {
      if (!world.isRemote) Block.blocksList(block).asInstanceOf[BlockCrops].fertilize(world, x, y, z)
      true
    }
    else {
      var i1 = 0
      var j1 = 0
      var k1 = 0
      if (block == Block.cocoaPlant.blockID) {
        i1 = world.getBlockMetadata(x, y, z)
        j1 = BlockDirectional.getDirection(i1)
        k1 = BlockCocoa.func_72219_c(i1)
        if (k1 >= 2) false
        else {
          if (!world.isRemote) {
            k1 += 1
            world.setBlockMetadataWithNotify(x, y, z, k1 << 2 | j1, 2)
          }
          true
        }
      }
      else if (block != Block.grass.blockID) false
      else {
        if (!world.isRemote) {
          for (i <- 0 until 128) {
            breakable {
              j1 = x
              k1 = y + 1
              var l1 = z
              val random = new Random()
              for (j <- 0 until (i / 16)) {
                j1 += random.nextInt(3) - 1
                k1 += (random.nextInt(3) - 1) * random.nextInt(3) / 2
                l1 += random.nextInt(3) - 1
                if (world.getBlockId(j1, k1 - 1, l1) != Block.grass.blockID || world.isBlockNormalCube(j1, k1, l1)) break()
              }
              if (world.getBlockId(j1, k1, l1) == 0) if (random.nextInt(10) != 0) if (Block.tallGrass.canBlockStay(world, j1, k1, l1)) world.setBlock(j1, k1, l1, Block.tallGrass.blockID, 1, 3)
              else ForgeHooks.plantGrass(world, j1, k1, l1)
            }
          }
        }
        true
      }
    }
    else if (world.getBlockMetadata(x, y, z) == 7) false
    else {
      if (!world.isRemote) Block.blocksList(block).asInstanceOf[BlockStem].fertilizeStem(world, x, y, z)
      true
    }
    else {
      if (!world.isRemote && world.rand.nextFloat.toDouble < 0.4D) Block.blocksList(block).asInstanceOf[BlockMushroom].fertilizeMushroom(world, x, y, z, world.rand)
      true
    }
  }

  def willTryCatalyze(world: World, x: Int, z: Int): Boolean = {
    val biome = world.getBiomeGenForCoords(x, z)
    val r = world.rand
    val temperature = biome.temperature
    val humidity = biome.rainfall
    possibilityMultiplier * (temperature * Math.abs(temperature) * (r.nextDouble - 0.3D) + humidity * Math.abs(humidity) * (r.nextDouble - 0.3D)) > 0.25D
  }
}
